### PR TITLE
Add proc_macro::Span::{before, after}.

### PR DESCRIPTION
**Pull request by m-ou-se**
This adds `proc_macro::Span::before()` and `proc_macro::Span::after()` to get a zero width span at the start or end of the span.

These are equivalent to rustc's `Span::shrink_to_lo()` and `Span::shrink_to_hi()` but with a less cryptic name. They are useful when generating diagnostlics like "missing \<thing\> after \<thing\>".

E.g.

```rust
syn::Error::new(ident.span().after(), "missing `:` after field name").into_compile_error()
```

### CODE REVIEW

File: library/proc_macro/src/lib.rs [hunk 1]
```diff
@@ -358,6 +358,18 @@ impl Span {
         self.0.end()
     }
 
+    /// Creates an empty span pointing to directly before this span.
+    #[unstable(feature = "proc_macro_span_shrink", issue = "none")]
```
Comment (m-ou-se) [inline library/proc_macro/src/lib.rs:362]: Note to self: This still needs a tracking issue.

**Review by petrochenkov — APPROVED** [review]

**Comment by jhpratt** [comment]
Part of me would prefer this be called `start` and `end`, but those names are already (unstably) taken.

Would it makes sense to repurpose these names and provide a generalized `.location()` method that would return the `LineColumn`? Presumably it could point to the start of the span. Alternatively `.start_location()` and `.end_location()` could be provided.

Just a random thought I had from a usability perspective; I could certainly get accustomed to the before/after naming.

**Comment by m-ou-se** [comment]
@jhpratt Since a span is not a point/location but a range, calling it start and end doesn't feel entirely accurate to me.

**Review by dtolnay — APPROVED** [review]
This API looks good to me as implemented.

**Pull request merged**